api/model.go

```diff
@@ -116,10 +116,19 @@ type EngineUpgradeInput struct {
 
 type Node struct {
 	client.Resource
-	Name            string          `json:"name"`
-	Address         string          `json:"address"`
-	AllowScheduling bool            `json:"allowScheduling"`
-	Status          types.NodeState `json:"status"`
+	Name            string              `json:"name"`
+	Address         string              `json:"address"`
+	AllowScheduling bool                `json:"allowScheduling"`
+	Disks           map[string]DiskInfo `json:"disks"`
+	Status          types.NodeState     `json:"status"`
+	AddDisks        []DiskInfo          `json:"addDisks"`
```

Comment:
This API is weird...

Previously we were using the path as the key, so it's easy to do CRUD inside node object since the caller knew the key anyway.

But now the key (`Filesystem ID`) for `Disks` won't be known for the caller until the disk has been created. So it's no longer proper to combine this information in the `Node` object.

Let's add an action to the `v1/nodes/<node>`, call it `diskUpdate`. And create `DiskUpdateInput` struct.

```
types DiskUpdateInput struct {
    []types.DiskSpec
}
```

During the processing, we can map `diskSpec.Path` to the existing `Disks` in the node, then compare to the `node.Disks` to know which to update, delete and create.

A little more about the update:
1. Update of `path` won't be allowed.
2. In the API, we can allow the user to specify the StorageMaximum value to override the system detected value, and shows a warning in the log with both values if the user chose to do so. This parameter can be useful for testing. But in the UI, the option shouldn't be shown for now.

Reply:
A little more about what's wrong with this API.

It's mainly because the API caller would need to do the comparison work for the backend.

Image the API shows a list to the user, and the user can add/modify/delete the item on the list freely. So how should UI interact with backend API?

There are two approaches in general:

**Action for each item**
It can be done when each item has been updated, and the user needs to click `Save` for every item for it to take effect.

But imagining the user wants to add 10 disks, it will be 10 extra clicks for the `Save` button. And each click will issue one separate API call to the backend, as well as each time the `Delete` button was clicked. And UI needs to refresh the screen every time this happened. Also, what if the user changes his/her mind during the editing process? The deleted items cannot be recalled since it's already gone in the backend.

So I think we can agree that this is a bad user experience for the user. That's why I don't want to have single item CRUD action on each of the disk.

BTW, this approach is fine with the Volumes. Those applies more to the configuration type of input, like disks or recurring jobs, since configuration are static. Volumes are independent objects, has operations of its own, also are more heavily weighted compared to the disks or recurring jobs.

**Action as a list**
It can also be done as UI only update the list when the user has done editing. Then UI will only need to issue one API call to the backend. That's easier.

But if the API was designed like in this code, UI would need to do the heavy-lifting of figuring out what's changed and put the result in the three catalogs the API specified.

And all those extra works UI does probably doesn't help much to reduce the backend work. For example, UI doesn't know if the disks `/mnt/disk1a` and `/mnt/disk1b` mounted to the same disk. Backend need to perform the validation work anyway.

So if we decided to accept the input as a list, let's just do the comparison work in the backend.